src/dataset.rs

```diff
@@ -148,31 +151,34 @@ pub fn filter_outlier(items: Vec<FSRSItem>) -> Vec<FSRSItem> {
 
     let mut filtered_items = vec![];
 
-    // 对每个按 rating 分组的子组进一步处理
-    for (_rating, delta_t_groups) in groups.iter() {
+    for (rating, delta_t_groups) in groups.iter() {
         let mut sub_groups = delta_t_groups.iter().collect::<Vec<_>>();
 
-        // 按子组大小升序排序，大小相同的按 delta_t 降序排序
+        // order by size of sub group ascending and delta_t descending
         sub_groups.sort_by(|(delta_t_a, subv_a), (delta_t_b, subv_b)| {
             subv_b
                 .len()
                 .cmp(&subv_a.len())
-                .then(delta_t_a.cmp(delta_t_b))
+                .then(delta_t_b.cmp(delta_t_a))
         });
 
-        // 计算总大小
+        // remove 5% of items from each sub group
         let total = sub_groups.iter().map(|(_, vec)| vec.len()).sum::<usize>();
         let mut has_been_removed = 0;
 
-        for (_delta_t, sub_group) in sub_groups.iter().rev() {
+        for (delta_t, sub_group) in sub_groups.iter().rev() {
             if has_been_removed + sub_group.len() > total / 20 {
                 filtered_items.extend_from_slice(sub_group);
             } else {
                 has_been_removed += sub_group.len();
+                // keep the items in trainset if they are not removed from filtered_items
+                trainset.retain(|item| {
+                    item.reviews[0].rating != *rating || item.reviews[1].delta_t != **delta_t
+                });
```

Comment:
.retain() is an expensive operation in a large vec, so we don't want to be doing this in a loop. It would be better to collect the removal criteria into a hashmap, then use that map to each each item in a single retain() call.